mypy/stubtest.py

```diff
@@ -1526,7 +1528,203 @@ def is_read_only_property(runtime: object) -> bool:
+        for name, default in zip(f.args.args, defaults):
+            p(name, default)
+
+    else:
+        args = reversed(f.args.args)
+        defaults = reversed(f.args.defaults)
+        iter = itertools.zip_longest(args, defaults, fillvalue=None)
+        if last_positional_only is not None:
+            kind = Parameter.POSITIONAL_ONLY
+        else:
+            kind = Parameter.POSITIONAL_OR_KEYWORD
+        for i, (name, default) in enumerate(reversed(list(iter))):
+            assert name is not None
+            p(name, default)
+            if i == last_positional_only:
+                kind = Parameter.POSITIONAL_OR_KEYWORD
+
+    # *args
+    if f.args.vararg:
+        kind = Parameter.VAR_POSITIONAL
+        p(f.args.vararg, empty)
+
+    # keyword-only arguments
+    kind = Parameter.KEYWORD_ONLY
+    for name, default in zip(f.args.kwonlyargs, f.args.kw_defaults):
+        p(name, default)
+
+    # **kwargs
+    if f.args.kwarg:
+        kind = Parameter.VAR_KEYWORD
+        p(f.args.kwarg, empty)
+
+    if self_parameter is not None:
+        # Possibly strip the bound argument:
+        #    - We *always* strip first bound argument if
+        #      it is a module.
+        #    - We don't strip first bound argument if
+        #      skip_bound_arg is False.
+        assert parameters
+        _self = getattr(obj, "__self__", None)
+        self_isbound = _self is not None
+        self_ismodule = inspect.ismodule(_self)
+        if self_isbound and (self_ismodule or skip_bound_arg):
+            parameters.pop(0)
+        else:
+            # for builtins, self parameter is always positional-only!
+            p = parameters[0].replace(kind=Parameter.POSITIONAL_ONLY)
+            parameters[0] = p
+
+    return cls(parameters, return_annotation=cls.empty)
+
+
 def safe_inspect_signature(runtime: Any) -> inspect.Signature | None:
+    if (
+        hasattr(runtime, "__name__")
+        and runtime.__name__ == "__init__"
+        and hasattr(runtime, "__text_signature__")
+        and runtime.__text_signature__ == "($self, /, *args, **kwargs)"
+        and hasattr(runtime, "__objclass__")
+        and runtime.__objclass__ is not object
+        and hasattr(runtime.__objclass__, "__text_signature__")
+        and runtime.__objclass__.__text_signature__ is not None
+    ):
+        # This is an __init__ method with the generic C-class signature.
+        # In this case, the underlying class usually has a better signature,
+        # which we can convert into an __init__ signature by adding $self
+        # at the start. If we hit an error, failover to the normal
+        # path without trying to recover.
+        if "/" in runtime.__objclass__.__text_signature__:
+            new_sig = f"($self, {runtime.__objclass__.__text_signature__[1:]}"
+        else:
+            new_sig = f"($self, /, {runtime.__objclass__.__text_signature__[1:]}"
+        try:
+            return _signature_fromstr(inspect.Signature, runtime, new_sig)
```

Comment:
Instead of parsing `__objclass__.__text_signature__`, would it be possible to transform `inspect.signature(runtime.__objclass__)`? Something like
```python
s = inspect.signature(runtime.__objclass__)
return s.replace(
    parameters=[
        inspect.Parameter("self", inspect.Parameter.POSITIONAL_OR_KEYWORD),
        *s.parameters.values(),
    ]
)
```

Reply:
ah, indeed. That's much better.